src/sage/rings/polynomial/polynomial_ring.py

```diff
@@ -1776,7 +1777,38 @@ def monics( self, of_degree=None, max_degree=None ):
         raise ValueError("you should pass exactly one of of_degree and max_degree")
 
 
-class PolynomialRing_commutative(PolynomialRing_general):
+# Placeholder class for deprecation
+class PolynomialRing_general(PolynomialRing_generic):
+    """
+    Univariate polynomial ring over a ring.
+
+    This class is deprecated. Please use :class:`PolynomialRing_generic`.
+    """
+
+    def __init__(self, *args, **kwds):
+        """
+        This class is deprecated. Please use :class:`PolynomialRing_generic`.
+
+        TESTS::
+
+            sage: from sage.rings.polynomial.polynomial_ring import PolynomialRing_general, PolynomialRing_generic
+            sage: PolynomialRing_general(QQ, name="a")
+            doctest:warning...DeprecationWarning:
+            The class PolynomialRing_general has been renamed to PolynomialRing_generic...
+            Univariate Polynomial Ring in a over Rational Field
+
+        Check they provide the same functionalities for backward compatability::
+
+            sage: dir(PolynomialRing_general(QQ, name="a")) == dir(PolynomialRing_generic(QQ, name="a"))
+            True
+        """
+        deprecation(
+            38207, "The class PolynomialRing_general has been renamed to PolynomialRing_generic."
+        )
+        super().__init__(*args, **kwds)
+
```

Comment:
Do we need this deprecation? `PolynomialRing_general` is an internal class.

Reply:
Technically yes because it is in the public documentation. However, this is such a core class that it is quite probable that code in the wild is using this in some form. In either case, it doesn’t really hurt to have this deprecation.

Reply:
I don't know, I am not familiar with deprecation of *classes*. Should there still be some kind of deprecation message for it? Or just let people git grep it? 😆

Reply:
Deprecated code needs another PR to remove it later. It is nontrivial maintenance burden. I don't think it is a good idea to add deprecation in sage based on "it doesn't really hurt".

This is just one of common cases, for which we should apply our principles. What is the principle of determining a class is "internal" or "public"? Is it public (1) if it appears in our reference manual or (2) if it is accessible in repl without explicitly importing the class? I was thinking it is (2), but you are saying it is (1). I am not sure. What others think?

Reply:
> I don't know, I am not familiar with deprecation of _classes_. 

I thought I knew. But I am not sure after hearing Travis :-) Let us hear Matthias too.

Reply:
Removing code is really basically nothing, and it isn’t something that needs to be done. So it is an epsilon maintenance burden at worst.

Strictly speaking, it is (1), but we don’t have to do that in all cases. In this case, we really should follow it.

Reply:
By the way if it is (2), we are perhaps destined to add lots of deprecations whenever we refactor classes...

Reply:
This question is separate from reviewing this PR.

I am positive to this PR, anyway.

Reply:
The doc https://deploy-livedoc--sagemath.netlify.app/html/en/developer/coding_in_python.html#deprecation seems to have (2) in mind.

Reply:
I don't see that interpretation at all. It only says "a backwards-incompatible change" with no reference to "internal" (or any synonym). The only time the global namespace is explicitly mentioned is for one example; note there is an "any other case" and even though it says nothing explicitly about classes, that is clearly implied. If we interpret it very strictly, then even hidden functions/methods/etc. are subject to deprecation. However, I would say for that if a user knows about such methods, then such a user can handle when such a change occurs. Moreover, users can easily see the online documentation, most will understand how to do imports, and they often subclass objects in their own code.

Reply:
Let's not read the developer guide as like legal statements. 

Deprecation policy is intended to guide users when they encounter changes while they use sage. First I thought only of using sage through user interfaces. But I now agree that we should include the cases when they code using the sage library. If a class could be used by user code and there is change in how to use it, then yes we should add deprecation warning somehow. I thought `PolynomialRing_general` is not such a class unlike `PolynomialRing`. But perhaps I am wrong in this. 

hidden functions/methods/classes are by definition not to used outside the sage library. Deprecation policy is not for those using them.

Reply:
I think that classes with an underscore in them may have been intended to be "internal" (after all, this naming scheme violates Python naming conventions).
But in the big `is_ClassName` deprecation, we just started to ask users to switch to using `isinstance(..., ClassName_generic)` in many cases. So that's now definitely public API

Reply:
Is someone who is writing `isinstance(..., ClassName_generic)` a sage user? Or is he a developer of the sage library?

Reply:
Could be either

Reply:
By the way, there is a discussion on similar topic: https://github.com/sympy/sympy/issues/23037. But I cannot find a good answer there either.

Reply:
> I think that classes with an underscore in them may have been intended to be "internal" (after all, this naming scheme violates Python naming conventions). 

This is a good idea for a criterion of "public" (being in API) or "internal" (being not part of API).

(C) A  name living in `sage.` is "internal" if it has an underscore in it, but "public" otherwise.

This is consistent with the policy that an attribute starting with an underscore is considered internal.

According to (C), `PolynomialRing_generic` is internal :-)

Reply:
> > I think that classes with an underscore in them may have been intended to be "internal" (after all, this naming scheme violates Python naming conventions).
> 
> This is a good idea for a criterion of "public" (being in API) or "internal" (being not part of API).
> 
> (C) A name living in `sage.` is "internal" if it has an underscore in it, but "public" otherwise.

Furthermore, all functions/methods with 2 words are therefore internal, such as `change_ring()`.

> This is consistent with the policy that an attribute starting with an underscore is considered internal.
>
> According to (C), `PolynomialRing_generic` is internal :-)

Of course, starting with an underscore was what was meant. :)

Reply:
No.

Reply:
We are talking about **class** names with an underscore anywhere in their name.

Reply:
Sorry. As we have already rules for functions/methods/attributes, the updated rule is

(C) Class names living in `sage.` is "internal" if it has an underscore anywhere in it, but "public" otherwise

Good enough?

Reply:
I am not entirely following the conversation, so maybe when you three reach a conclusion (and maybe put it somewhere in the developer guide) I can modify this PR accordingly. I'm fine with any solution, I came up with this placeholder class knowing it's ad-hoc-y

Reply:
(And ping me for it :) )

Reply:
The general solution being discussed in this thread is somewhat separate from reviewing this PR.

Whether or not we reach a conclusion for the general problem here, I am willing to continue this discussion in sage-devel. If the community reach to a solution (either by consensus or by voting), yes we should put it to the developer guide.

You, as the author, do not have to wait for that, but just resolve the work issue below according to the reviewer's request (which may change by the discussion here, of course).

Reply:
> You, as the author, do not have to wait for that, but just resolve the work issue below according to the reviewer's request (which may change by the discussion here, of course).

Well, Matthias' review/suggestion would not be necessary if the conclusion in this discussion is that this class is treated as internal, since it then doesn't have to go through the deprecation process. Plus, as I replied to Matthias below, I don't agree with his proposed change, and he hasn't replied to me or Travis yet.